### PR TITLE
Production Readiness Checklist

### PR DESCRIPTION
**Pull request by brikis98**
Add a “Production Readiness Checklist” to our website, which is a checklist of everything you need to do before going live on AWS. I’ll paste a screenshot of what it looks like shortly.

* All the items are defined in a YAML file.
* The items are broken down by categories: server-side, client-side, data storage, etc.
* Each item has a description. To keep the page from getting enormous, only the first sentence of the description is shown by default, but you can click a "more" link to see it all.
* The page uses local storage to remember what items you've checked.

I'd love to launch this on Monday with a small blog post.

### CODE REVIEW

**Comment by brikis98** [comment]
Here's what it looks like:

screencapture-localhost-4000-devops-checklist-2018-05-14-02_11_07

It's a long page, as there's a lot to do!

File: _data/devops-checklist.yml [hunk 1]
```diff
@@ -0,0 +1,690 @@
+- category: Server-side
+  items:
+    - title: Build AMIs
+      description: |
+        If you want to run your apps directly on EC2 Instances, you should package them as
+        <a href="https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/AMIs.html" target="_blank">Amazon Machine Images (AMIs)</a>
+        using a tool such as <a href="https://www.packer.io/" target="_blank">Packer</a>. Although we recommend Docker
+        for all stateful apps (see below), we recommend directly using AMIs and EC2 Instances for all stateful apps,
```
Comment (eak12913) [inline _data/devops-checklist.yml]: for all ~stateful~ stateless apps (see below),...

File: _data/devops-checklist.yml [hunk 2]
```diff
@@ -0,0 +1,690 @@
...
+        cluster of EC2 Instances, and Amazon takes care of scheduling containers across them. Another is
+        <a href="https://aws.amazon.com/eks/" target="_blank">Elasic Kubernetes Service (EKS)</a>, which is a Amazon's
+        managed <a href="https://kubernetes.io/" target="_blank">Kubernetes</a> (note, EKS is still in preview mode
+        as of May, 2018). A third option is <a href="https://aws.amazon.com/fargate/" target="_blank">AWS Fargate</a>,
+        a service where AWS manages and scales the underlying Instances for you and you just hand it Docker containers
+        to run.
+
+    - title: Deploy serverless apps using Lambda and API Gateway
+      description: |
+        If want to build serverless apps, you should package them as
```
Comment (eak12913) [inline _data/devops-checklist.yml]: If _you_ want to build...

File: _data/devops-checklist.yml [hunk 3]
```diff
@@ -0,0 +1,690 @@
...
+
+    - title: Deploy Docker images using ECS, EKS, or Fargate
+      description: |
+        You have several options for running Docker containers in AWS. One is to use the
+        <a href="https://aws.amazon.com/ecs/" target="_blank">Elastic Container Service (ECS)</a>, where you run a
+        cluster of EC2 Instances, and Amazon takes care of scheduling containers across them. Another is
+        <a href="https://aws.amazon.com/eks/" target="_blank">Elasic Kubernetes Service (EKS)</a>, which is a Amazon's
+        managed <a href="https://kubernetes.io/" target="_blank">Kubernetes</a> (note, EKS is still in preview mode
+        as of May, 2018). A third option is <a href="https://aws.amazon.com/fargate/" target="_blank">AWS Fargate</a>,
+        a service where AWS manages and scales the underlying Instances for you and you just hand it Docker containers
```
Comment (eak12913) [inline _data/devops-checklist.yml]: Instances -> instances

File: _data/devops-checklist.yml [hunk 4]
```diff
@@ -0,0 +1,690 @@
...
+    - title: Deploy serverless apps using Lambda and API Gateway
+      description: |
+        If want to build serverless apps, you should package them as
+        <a href="https://docs.aws.amazon.com/lambda/latest/dg/deployment-package-v2.html" target="_blank">deployment packages</a>
+        for <a href="https://aws.amazon.com/lambda/" target="_blank">AWS Lambda</a>. You can expose your Lambda
+        functions as HTTP endpoints using <a href="https://aws.amazon.com/api-gateway/" target="_blank">API Gateway</a>.
+
+    - title: Configure CPU, memory, and GC settings
+      description: |
+        Configure memory settings CPU settings (e.g., ensure your app uses all available CPU cores using tools such
```
Comment (eak12913) [inline _data/devops-checklist.yml]: Configure ~memory settings~ CPU settings

File: _data/devops-checklist.yml [hunk 5]
```diff
@@ -0,0 +1,690 @@
...
+        as Node Cluster), memory settings (e.g., <code>-Xmx, -Xms</code> settings for the JVM), and GC settings (if
+        applicable) for your app. If you're deploying directly on EC2 Instances, these should be configured based on
+        the available CPU and memory on your Instance (see
```
Comment (eak12913) [inline _data/devops-checklist.yml]: It seems you've purposely capitalized `Instances` when you refer to EC2 instances.  I'm not entirely sure it should be capitalized, especially when you just say something like "...on your Instance...". Seems like it should be `instance` there.

File: pages/devops-checklist/_sub-hero.html [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+    <p>
+      This checklist is your guide to the best practices for deploying secure, scalable, and highly available
+      infrastructure in AWS. Before you go live, go through each item, and make sure you haven't missed anything
+      important!
+    </p>
+    <ol>
+      {% for category in site.data.devops-checklist %}
+        <li><a href="#{{ category.category | slugify }}">{{ category.category}}</a></li>
+      {% endfor %}
+    </ol>
```
Comment (josh-padnick) [inline pages/devops-checklist/_sub-hero.html]: Consider adding a little note informing users that checked boxes will persist in local storage across browser sessions.

File: _data/devops-checklist.yml [hunk 7]
```diff
@@ -0,0 +1,690 @@
...
+    - title: Use a static content server
+      description: |
+        You should serve all your static content (CSS, JS, images, fonts) from a static content server so that your
+        dynamic web framework (e.g., from Rails, Node.js, or Django) can focus solely on processing dynamic requests.
+        The best static content host to use with AWS is <a href="https://aws.amazon.com/s3/" target="_blank">S3</a>.
+
+    - title: Use a CDN
+      description: |
+        Use <a href="https://aws.amazon.com/cloudfront/" target="_blank">CloudFront</a> as a
+        <a href="https://en.wikipedia.org/wiki/Content_delivery_network" target="_blank">Content Distribution Network (CDN)</a>
```
Comment (eak12913) [inline _data/devops-checklist.yml]: Delivery vs distribution?

File: _data/devops-checklist.yml [hunk 8]
```diff
@@ -0,0 +1,690 @@
...
+        <a href="https://aws.amazon.com/rds/aurora/" target="_blank">Amazon Aurora</a> as a highly scalable,
+        cloud-native, MySQL and PostgreSQL compatible database. Both RDS and Aurora support automatic failover, read
+        replicas, and automated backup.
+
+    - title: Deploy NoSQL databases
+      description: |
+        Use <a href="https://aws.amazon.com/elasticache/" target="_blank">Elasticache</a> if you want to use Redis or
+        Memcached for key-value storage. Use <a href="https://aws.amazon.com/dynamodb/" target="_blank">DynamoDB</a> if
+        you need a managed, eventually consistent document store. If you need other NoSQL databases, such as MongoDB
+        or Couchbase, you'll need to run it yoruself (see the
```
Comment (eak12913) [inline _data/devops-checklist.yml]: run it ~yoruself~ yourself...

File: _data/devops-checklist.yml [hunk 9]
```diff
@@ -0,0 +1,690 @@
...
+        Set up a build system for your project, such as <a href="https://gradle.org/" target="_blank">Gradle</a> (for
+        Java), <a href="https://github.com/ruby/rake" target="_blank">Rake</a> (for Ruby), or
+        <a href="https://yarnpkg.com/en/" target="_blank">Yarn</a> (for Node.js). The build system is responsible for
+        compiling your app, as well as many other tasks described below.
+
+    - title: Use dependency management
+      description: |
+        Your build systems should allow you to explicitly define all the of the dependencies for your apps. Each
+        dependency should be versioned, and ideally, the versions of all dependencies, including transitive
+        dependencies, are captured in a lock file.
```
Comment (eak12913) [inline _data/devops-checklist.yml]: May need some more detail around the "lock" file or a link where you can read more about it.

File: _data/devops-checklist.yml [hunk 10]
```diff
@@ -0,0 +1,690 @@
...
+        Deploy immutable artifacts to one environment at a time, and promote it to the next environment after testing.
+        For example, you might deploy v0.3.2 to dev, and test it there. If it works well, you promote the exact same
+        artifact, v0.3.2, to stage, and test it there. If all goes well, you finally promote v0.3.2 to prod. Since it's
+        the exact same code in every environment, there's a good chance that if it works in one environment, it'll
+        also work in the others.
+
+    - title: Roll back in case of failure
+      description: |
+        If you use immutable, versioned artifacts as your unit of deployment, then any time something goes wrong, you
+        have the option to roll back to a known-good state by rolling out the previous version. If your infrastructure
```
Comment (eak12913) [inline _data/devops-checklist.yml]: by rolling _back_ the previous version?

File: _data/devops-checklist.yml [hunk 11]
```diff
@@ -0,0 +1,690 @@
...
+
+    - title: Configure Security Groups
+      description: |
+        Every AWS resource (e.g., EC2 Instances, Load Balancers, RDS DBs, etc.) has a
+        <a href="https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-network-security.html" target="_blank">Security Group</a>
+        that acts as a firewall, controlling what traffic is allowed in and out of that resource. By default, no
+        traffic is allowed in or out. Follow the
+        <a href="https://en.wikipedia.org/wiki/Principle_of_least_privilege" target="_blank">Principle of Least Privilege</a>
+        and open up the absolute minimum number of ports you can for each resource. When opening up a port, you can
+        also specify either the CIDR block (IP address range) or ID of another Security Group that are allowed to
```
Comment (eak12913) [inline _data/devops-checklist.yml]: `or ID of another Security Group that are allowed to` -> ...that _is_ allowed to...

File: _data/devops-checklist.yml (hunk 1 above)
Comment (brikis98) [reply _data/devops-checklist.yml]: Fixed! https://github.com/gruntwork-io/gruntwork-io.github.io/pull/45/commits/ac969266fa45512affe5434f70111b488f8c4eb8

File: _data/devops-checklist.yml (hunk 2 above)
Comment (brikis98) [reply _data/devops-checklist.yml]: Fixed! https://github.com/gruntwork-io/gruntwork-io.github.io/pull/45/commits/ac969266fa45512affe5434f70111b488f8c4eb8

File: _data/devops-checklist.yml (hunk 3 above)
Comment (brikis98) [reply _data/devops-checklist.yml]: See below

File: _data/devops-checklist.yml (hunk 4 above)
Comment (brikis98) [reply _data/devops-checklist.yml]: Fixed! https://github.com/gruntwork-io/gruntwork-io.github.io/pull/45/commits/ac969266fa45512affe5434f70111b488f8c4eb8

File: _data/devops-checklist.yml (hunk 5 above)
Comment (brikis98) [reply _data/devops-checklist.yml]: Yea, we try to capitalize AWS Terminology: EC2 Instance, Security Group, etc. Not sure what the "proper" way is, but it's what we've been doing so far.

File: _data/devops-checklist.yml (hunk 7 above)
Comment (brikis98) [reply _data/devops-checklist.yml]: Both are accepted (wiki page list both).

File: _data/devops-checklist.yml (hunk 8 above)
Comment (brikis98) [reply _data/devops-checklist.yml]: Fixed! https://github.com/gruntwork-io/gruntwork-io.github.io/pull/45/commits/ac969266fa45512affe5434f70111b488f8c4eb8

File: _data/devops-checklist.yml [hunk 12]
```diff
@@ -0,0 +1,690 @@
...
+        check those files into version control, and apply the migrations as part of the deployment process. See
+        <a href="https://flywaydb.org/" target="_blank">Flyway</a> and
+        <a href="https://www.liquibase.org/" target="_blank">Liquibase</a>.
+
+- category: Scalability and High Availability
+  items:
+    - title: Choose between a Monolith and Microservices
+      description: |
+        Ignore the hype and stick with a monolithic architecture as long as you possibly can. Microservices have a
+        massive costs (operational overhead, performance overhead, more failure modes, loss of
```
[the remark this answers is not in the document]
Comment (brikis98) [reply _data/devops-checklist.yml]: Fixed! https://github.com/gruntwork-io/gruntwork-io.github.io/pull/45/commits/ac969266fa45512affe5434f70111b488f8c4eb8

File: _data/devops-checklist.yml (hunk 9 above)
Comment (brikis98) [reply _data/devops-checklist.yml]: Good point. Added a couple links. https://github.com/gruntwork-io/gruntwork-io.github.io/pull/45/commits/ac969266fa45512affe5434f70111b488f8c4eb8

File: _data/devops-checklist.yml (hunk 10 above)
Comment (brikis98) [reply _data/devops-checklist.yml]: Fixed! https://github.com/gruntwork-io/gruntwork-io.github.io/pull/45/commits/ac969266fa45512affe5434f70111b488f8c4eb8

File: _data/devops-checklist.yml (hunk 11 above)
Comment (brikis98) [reply _data/devops-checklist.yml]: Fixed! https://github.com/gruntwork-io/gruntwork-io.github.io/pull/45/commits/ac969266fa45512affe5434f70111b488f8c4eb8

File: _data/devops-checklist.yml (hunk 5 above)
Comment (eak12913) [reply _data/devops-checklist.yml]: Makes sense!

File: pages/devops-checklist/_sub-hero.html (hunk 6 above)
Comment (brikis98) [reply pages/devops-checklist/_sub-hero.html]: Done: https://github.com/gruntwork-io/gruntwork-io.github.io/pull/45/commits/fed86ece7bbdf16a1fe88220450df7eec8a94c60. I also added "check all", "uncheck all", etc controls in that commit.

**Comment by brikis98** [comment]
Thx for the feedback everyone. Merging now.

**Pull request merged**